Sources/BraveWallet/DecentralizedDNSHelper.swift

```diff
@@ -0,0 +1,99 @@
+// Copyright 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import BraveCore
+import Shared
+
+public enum Web3Service: String, CaseIterable {
+  case solana
+  case ethereum
+  case ethereumOffchain
+  
+  public var id: String { rawValue }
+}
+
+@MainActor public class DecentralizedDNSHelper {
+  
+  private let rpcService: BraveWalletJsonRpcService
+  private let ipfsApi: IpfsAPI?
+  
+  public init?(
+    rpcService: BraveWalletJsonRpcService,
+    ipfsApi: IpfsAPI?,
+    isPrivateMode: Bool
+  ) {
+    guard !isPrivateMode else { return nil }
+    self.rpcService = rpcService
+    self.ipfsApi = ipfsApi
+  }
+  
+  public enum DNSLookupResult: Equatable {
+    case none
+    case loadInterstitial(Web3Service)
+    case load(URL)
+  }
+  
+  public func lookup(domain: String) async -> DNSLookupResult {
+    guard let fixupURL = URIFixup.getURL(domain) else {
+      return .none
+    }
+    if fixupURL.domainURL.schemelessAbsoluteDisplayString.endsWithSupportedENSExtension {
+      return await lookupENS(domain: domain)
+    } else if fixupURL.domainURL.schemelessAbsoluteDisplayString.endsWithSupportedSNSExtension {
+      return await lookupSNS(domain: domain)
+    }
+    return .none
+  }
+  
+  /// Decentralized DNS lookup for an ENS domain
+  private func lookupENS(domain: String) async -> DNSLookupResult {
+    let ensResolveMethod = await rpcService.ensResolveMethod()
+    switch ensResolveMethod {
+    case .ask:
+      return .loadInterstitial(.ethereum)
+    case .enabled:
+      let (contentHash, isOffchainConsentRequired, status, _) = await rpcService.ensGetContentHash(domain)
+      if isOffchainConsentRequired {
+        return .loadInterstitial(.ethereumOffchain)
+      }
+      guard status == .success,
+            !contentHash.isEmpty,
+            let ipfsUrl = ipfsApi?.contentHashToCIDv1URL(for: contentHash),
```

Comment:
The [objective-c wrapper](https://github.com/brave/brave-core/blob/2e55cbbe62ef7c9b2430ab2e24e1fb26d17ac1b5/ios/browser/api/ipfs/ipfs_api.h#L25) for this function  uses [NSURLWithGURL](https://source.chromium.org/chromium/chromium/src/+/main:net/base/mac/url_conversions.h;l=15-19;drc=5abffc9bbab0296c39d03cc4b894388a0b7f3258) which will return `nil` in Swift for invalid urls (ex. `GURL()`) so it would fail this guard check returning `DNSLookupResult.none`